src/libp2p/peer_id.rs

```diff
@@ -37,39 +33,68 @@ pub enum PublicKey {
 }
 
 impl PublicKey {
-    /// Encode the public key into a Protobuf structure for storage or
-    /// exchange with other nodes.
+    /// Encode the public key into a protobuf structure for exchange with other nodes.
+    ///
+    /// As indicated in the libp2p specification, the encoding is done deterministically despite
+    /// the fact that the protobuf format isn't deterministic.
+    ///
+    /// See https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#keys.
     pub fn to_protobuf_encoding(&self) -> Vec<u8> {
-        let public_key = match self {
-            PublicKey::Ed25519(key) => keys_proto::PublicKey {
-                r#type: keys_proto::KeyType::Ed25519 as i32,
-                data: key.to_vec(),
-            },
-        };
-
-        let mut buf = Vec::with_capacity(public_key.encoded_len());
-        public_key.encode(&mut buf).unwrap();
-        buf
+        match self {
+            PublicKey::Ed25519(key) => {
+                const CAPACITY: usize = 32 + 4;
+                let mut out = Vec::with_capacity(CAPACITY);
+                for slice in protobuf::enum_tag_encode(1, 1) {
+                    out.extend_from_slice(slice.as_ref());
+                }
+                for slice in protobuf::bytes_tag_encode(2, key) {
+                    out.extend_from_slice(slice.as_ref());
+                }
+                debug_assert_eq!(out.len(), CAPACITY);
```

Comment:
So it's a bit complicated. The expected length of the final buffer is not so easily calculable because numbers are encoded in LEB128, in other words in a variable-length way.
For example, if you send a buffer of 30 bytes, the message will be 32 bytes (30 + 2). But if you send a buffer of 150 bytes, the message will be 153 bytes (150 + 3).

I've done this `debug_assert_eq!` thing in `peer_id.rs` because the encoding of a PeerId is fully deterministic, but in other places I feel like hardcoding a capacity would be a bit too hacky.